engine/src/main/java/org/terasology/rendering/openvrprovider/OpenVRProvider.java

```diff
@@ -163,12 +189,87 @@ public void shutdown() {
         initialized = false;
     }
 
+    /**
+     * Query the VR library and update the VR state, gettable with getState(). This method should be called once per
+     * render cycle.
+     */
     public void updateState() {
         updatePose();
         pollControllers();
         pollInputEvents();
     }
 
+    /**
+     * Make the specified controller vibrate
+     * @param controller - the hand index, 0 for left and 1 for right.
+     * @param strength - the strength of the pulse.
+     */
+    public static void triggerHapticPulse(int controller, int strength) {
+        if (controllerDeviceIndex[controller] == -1) {
+            return;
+        }
+        vrSystem.TriggerHapticPulse.apply(controllerDeviceIndex[controller], 0, (short) strength);
+    }
+
+    public void submitFrame() {
+        for (int nEye = 0; nEye < 2; nEye++) {
+            vrCompositor.Submit.apply(
+                    nEye,
+                    texType[nEye], null,
+                    JOpenVRLibrary.EVRSubmitFlags.EVRSubmitFlags_Submit_Default);
+        }
+        if (vrCompositor.PostPresentHandoff != null) {
+            vrCompositor.PostPresentHandoff.apply();
+        }
+    }
+
+    /**
+     * Set the near clip. This will influence the projection matrix that comes from the
+     * vrProvider.getState().getProjectionMatrix(...) method.
+     * @param nearClipIn - the near clip to set.
```

Comment:
For VR head sets, it's very important to use the projection matrix / pose matrices given by the SDK when rendering. I've never tried to put a negative value in, but an orthographic projection in general is a no-go for VR. The reason is that, if you draw a ray from the user's left eye to the apparent location of the object in 3D space, and then draw a line from that object back to the user's right eye, the angle that ray enters the user's eye had better agree with where the left eye is perceiving the object to be. If not, the user will get double vision and headaches.